File: Hash Table/219_Contains-Duplicate-II/219_Contains Duplicate II.cpp
/*******************************************************************************
 * Title:
 *     219. Contains Duplicate II
 *     219. 存在重复元素 II
 * Address:
 *     https://leetcode-cn.com/problems/contains-duplicate-ii/
 ******************************************************************************/

#include <iostream>
#include <vector>
#include <set>
using namespace std;

// 方法一：哈希表
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        set<int> record;
        for ( int i = 0; i < (int)nums.size(); ++i ) {
            if ( record.find( nums[i] ) != record.end() )
                return true;
            
            record.insert( nums[i] );
            if ( record.size() > k )
                record.erase( nums[i - k] );
        }

        return false;
    }
};

int main()
{
    

    return 0;
}

File: Hash Table/219_Contains-Duplicate-II/219_Contains Duplicate II.py
"""
 * Title:
 *     219. Contains Duplicate II
 *     219. 存在重复元素 II
 * Address:
 *     https://leetcode-cn.com/problems/contains-duplicate-ii/
"""

class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        record = set()
        for i in range(len(nums)):
            if nums[i] in record:
                return True
            
            record.add( nums[i] )
            if (len(record) > k):
                record.discard(nums[i - k])
        
        return False